python-frontend/src/main/java/org/sonar/python/tree/BinaryExpressionImpl.java

```diff
@@ -103,4 +107,17 @@ public Kind getKind() {
   public List<Tree> computeChildren() {
     return Stream.of(leftOperand, operator, rightOperand).filter(Objects::nonNull).collect(Collectors.toList());
   }
+
+  @Override
+  public InferredType type() {
+    if (is(Kind.AND, Kind.OR)) {
+      return InferredTypes.or(leftOperand.type(), rightOperand.type());
+    }
+    return InferredTypes.anyType();
+  }
+
+  @Override
+  public List<Expression> typeDependencies() {
+    return Arrays.asList(leftOperand, rightOperand);
```

Comment:
To avoid unnecessary calculations, it would be better to return an empty list when `!is(Kind.AND, Kind.OR)`.